### PR TITLE
feat: schema contracts

### PR DESCRIPTION
**Pull request by thisisnithin**
## Motivation and Context

https://github.com/wundergraph/cosmo/assets/44922285/42e12c22-7bba-492b-9fac-9e6ee2e9f704

## TODO

- [x] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/test/contracts.test.ts [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+      name: contractGraphName,
+      namespace: prod,
+    });
+    expect(getContractRes.response?.code).toEqual(EnumStatusCode.OK)
+    expect(getContractRes.subgraphs.length).toEqual(1)
+
+    await server.close();
+  });
+
+  test('Contract is deleted on deleting source monograph', async (testContext) => {
```
Comment (StarpTech) [inline controlplane/test/contracts.test.ts:402]: Please also test if the artifacts are deleted.

File: controlplane/test/contracts.test.ts [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+    });
+    expect(createResp.response?.code).toBe(EnumStatusCode.OK);
+
+    await client.createContract({
+      name: contractGraphName,
+      namespace: 'default',
+      sourceGraphName: monographName,
+      excludeTags: ['test'],
+      routingUrl: 'http://localhost:8081',
+      admissionWebhookUrl: 'http://localhost:8085',
```
Comment (StarpTech) [inline controlplane/test/contracts.test.ts:461]: Did you test that admission config is optional?

File: demo/pkg/subgraphs/employees/subgraph/schema.graphqls [hunk 3]
```diff
@@ -1,142 +1,143 @@
 directive @goField(
-    forceResolver: Boolean
```
Comment (StarpTech) [inline demo/pkg/subgraphs/employees/subgraph/schema.graphqls:2]: Just formatting or did you change anything?

File: shared/test/router.config.test.ts [hunk 4]
```diff
@@ -130,7 +130,7 @@ describe('Router Config Builder', () => {
     };
     const routerConfig = buildRouterConfig({
       // if the federatedClientSDL is empty, it is not added to the config
-      federatedClientSDL: federationResult.shouldIncludeClientSchema
+      federatedClientSDL: federationResult!.shouldIncludeClientSchema
```
Comment (StarpTech) [inline shared/test/router.config.test.ts:133]: Why can it be optional? Can we avoid `!`?

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Comment by StarpTech** [comment]
Feedback regarding UI: Why should the composed version be the default option in the UI? Logically, I'm interested in the version that is delivered to my users. Additionally, how can we explain the client and composed schema clearly to users?

File: demo/pkg/subgraphs/employees/subgraph/schema.graphqls (hunk 3 above)
Comment (thisisnithin) [reply demo/pkg/subgraphs/employees/subgraph/schema.graphqls:2]: New query added and implemented, `firstEmployee`.

File: controlplane/test/contracts.test.ts (hunk 2 above)
Comment (thisisnithin) [reply controlplane/test/contracts.test.ts:461]: It is optional. Some tests I do not pass it

File: cli/src/commands/contract/commands/create.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  const command = new Command('create');
+  command.description('Creates a contract graph from the provided source graph.');
+  command.argument('<name>', 'The name of the contract graph to create.');
+  command.option('-n, --namespace [string]', 'The namespace of the contract graph.');
+  command.requiredOption('--source [string]', 'The source graph from which you want to create the contract.');
+  command.requiredOption(
+    '-r, --routing-url <url>',
+    'The routing url of your router. This is the url that the router will be accessible at.',
+  );
+  command.option('--exclude [tags...]', 'Schema elements with these tags will be excluded from the contract schema.');
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply cli/src/commands/contract/commands/create.ts:22]: Yes

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 6]
```diff
@@ -1314,26 +1418,381 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+          return {
+            response: {
+              code: EnumStatusCode.ERR_NOT_FOUND,
+              details: `The graph is not a contract`,
+            },
+            compositionErrors: [],
+            deploymentErrors: [],
+          };
+        }
+
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:1730]: This is only for tags. Everything else is using federated graph commands

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 7]
```diff
@@ -1314,26 +1418,381 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+            response: {
+              code: EnumStatusCode.ERR,
+              details: `Provided tags are invalid`,
+            },
+            compositionErrors: [],
+            deploymentErrors: [],
+          };
+        }
+
+        if (!authContext.hasWriteAccess) {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:1697]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 8]
```diff
@@ -2860,27 +3311,50 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
             federatedGraphTargetId: graph.targetId,
           });
 
+          const deletedGraphs: FederatedGraphDTO[] = [];
+
+          // Delete downstream contracts
+          const contracts = await contractRepo.bySourceFederatedGraphId(graph.id);
+          for (const contract of contracts) {
+            const contractGraph = await fedGraphRepo.byId(contract.downstreamFederatedGraphId);
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:3319]: Ah right. Moved it to after all graphs are successfully deleted.

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 9]
```diff
@@ -530,8 +545,20 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
           authContext,
         });
 
+        const movedGraphs = [graph];
         const targetIds = [graph.targetId];
 
+        const contracts = await contractRepo.bySourceFederatedGraphId(graph.id);
```
Comment (StarpTech) [inline controlplane/src/core/bufservices/PlatformService.ts:551]: Please add a comment about this.

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 10]
```diff
@@ -1649,47 +2109,48 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+          if (composition.errors.length > 0 || !isInspectable || inspectorChanges.length === 0) {
+            continue;
+          }
 
-            const result = await trafficInspector.inspect(inspectorChanges, {
-              daysToConsider: limit,
-              federatedGraphId: composition.id,
-              organizationId: authContext.organizationId,
-            });
+          if (limit <= 0) {
```
Comment (StarpTech) [inline controlplane/src/core/bufservices/PlatformService.ts:2126]: Do we really need it here? If not, remove.

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 11]
```diff
@@ -2860,27 +3311,47 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
             federatedGraphTargetId: graph.targetId,
           });
 
-          const blobStorageDirectory = `${authContext.organizationId}/${graph.id}`;
-          await opts.blobStorage.removeDirectory({ key: blobStorageDirectory });
+          const deletedGraphs: FederatedGraphDTO[] = [];
+
+          // Delete downstream contracts
+          const contracts = await contractRepo.bySourceFederatedGraphId(graph.id);
```
Comment (StarpTech) [inline controlplane/src/core/bufservices/PlatformService.ts:3317]: Put the logic in a service method.

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 12]
```diff
@@ -2860,27 +3311,47 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+            deletedGraphs.push(contractGraph);
+          }
 
           await fedGraphRepo.delete(graph.targetId);
 
+          deletedGraphs.unshift(graph);
+
+          for (const deletedGraph of deletedGraphs) {
+            const blobStorageDirectory = `${authContext.organizationId}/${deletedGraph.id}`;
+            await opts.blobStorage.removeDirectory({ key: blobStorageDirectory });
```
Comment (StarpTech) [inline controlplane/src/core/bufservices/PlatformService.ts:3334]: Put a try/catch around and log the error so we get a notification when data could not be deleted.

File: controlplane/src/core/bufservices/PlatformService.ts (hunk 11 above)
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:3305]: Done

File: controlplane/src/core/bufservices/PlatformService.ts (hunk 10 above)
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Removed

File: controlplane/src/core/repositories/SubgraphRepository.ts [hunk 13]
```diff
@@ -447,6 +554,38 @@ export class SubgraphRepository {
         });
 
         deploymentErrors.push(...deployment.errors);
+
+        if (!federationResultContainerByContractName) {
+          continue;
+        }
+
+        for (const [contractName, resultContainer] of federationResultContainerByContractName.entries()) {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/repositories/SubgraphRepository.ts]: Done

File: controlplane/src/core/repositories/SubgraphRepository.ts [hunk 14]
```diff
@@ -350,6 +386,49 @@ export class SubgraphRepository {
               message: e.message ?? '',
             })),
         );
+
+        if (!federationResultContainerByContractName) {
+          continue;
+        }
+
+        for (const [contractName, resultContainer] of federationResultContainerByContractName.entries()) {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/repositories/SubgraphRepository.ts]: Done

File: controlplane/src/core/composition/composer.ts [hunk 15]
```diff
@@ -345,27 +372,56 @@ export class Composer {
...
-          namespaceId: graph.namespaceId,
-          targetID: graph.targetId,
-          fieldConfigurations: result?.fieldConfigurations || [],
-          composedSchema: result?.federatedGraphSchema
-            ? printSchemaWithDirectives(result.federatedGraphSchema)
-            : undefined,
-          errors: errors || [],
-          subgraphs: subgraphDTOsToComposedSubgraphs(subgraphs, result),
-        });
+        const contracts = await this.contractRepo.bySourceFederatedGraphId(graph.id);
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/composition/composer.ts:389]: This needs to be here to pass to the compose method

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 16]
```diff
@@ -3105,18 +3636,78 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+                errors: (errors || []).length > 0 || deploymentErrors.length > 0,
                 actor_id: authContext.userId,
               },
             });
+
+            if (!federationResultContainerByContractName) {
+              continue;
+            }
+
+            for (const [contractName, resultContainer] of federationResultContainerByContractName.entries()) {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 17]
```diff
@@ -3063,24 +3564,54 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
...
+              continue;
+            }
 
-            const namespace = await namespaceRepo.byTargetId(composition.targetID);
+            const namespace = await namespaceRepo.byTargetId(federatedGraph.targetId);
             if (!namespace) {
               throw new PublicError(EnumStatusCode.ERR_NOT_FOUND, 'Could not find namespace');
             }
 
+            const subgraphs = await subgraphRepo.listByFederatedGraph({
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Moved into federated graph repo

File: controlplane/src/core/bufservices/PlatformService.ts [hunk 18]
```diff
@@ -3063,24 +3564,54 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
 
           // Validate all federated graphs that use this subgraph.
           for (const federatedGraph of affectedFederatedGraphs) {
-            const composition = await composer.composeFederatedGraph(federatedGraph);
+            // only execute for base graphs. Contracts for each are handled automatically.
+            if (federatedGraph.contract) {
```
[the remark this answers is not in the document]
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts]: Done

File: controlplane/src/core/bufservices/PlatformService.ts (hunk 9 above)
Comment (thisisnithin) [reply controlplane/src/core/bufservices/PlatformService.ts:553]: Done

**Review by StarpTech — APPROVED** [review]
Well done!

**Pull request merged**